cohttp/test/unitary_test_header.ml

```diff
@@ -0,0 +1,372 @@
+  let h2 = H.(add (remove h "second") "second" "2a") in
+  aeh "update_all existing header" h1 h2;
+  let h1 = H.update_all h "second" (function [] -> [ "3" ] | _ -> []) in
+  let h2 = H.remove h "second" in
+  aeh "update_all remove header" h1 h2;
+  let h1 = H.update_all h "accept" (function [] -> [] | _ -> [ "baz" ]) in
+  aesl "update_all existing header with multiple values"
+    H.(get_multi h1 "accept")
+    [ "baz" ];
+  let h1 =
+    H.update_all h "accept" (function [] -> [] | xs -> xs @ [ "baz" ])
+  in
+  let h2 = H.add h "accept" "baz" in
+  aeso "update_all_existing_header_multivalued"
+    (H.get_multi_concat h1 "accept")
+    (H.get_multi_concat h2 "accept");
+  let h1 = H.update_all h "accept" (function _ -> []) in
+  aeh "update_all_existing_header_multivalue : remove all" (H.remove h "accept")
+    h1;
+  let h1 = H.update_all h "third" (function [] -> [ "3"; "33" ] | _ -> []) in
+  let h2 = H.add_multi h "third" [ "3"; "33" ] in
+  aeh "update add new header" h1 h2;
+  let h1 = H.update_all h "third" (function _ -> []) in
+  aeh "update_remove_absent_header" h h1;
+  let h1 = H.update_all h "third" (function [] -> [] | _ -> [ "3" ]) in
+  aeh "update_new_header: unchanged" h h1
+
+let get_multi_tests () =
+  aesl "get_multi (init ()) k" [] H.(get_multi (init ()) "a");
+  aesl "get_multi h k when mem h k = false" [] H.(get_multi prebuilt "a");
+  aesl "get_multi h k when mem h k = true" [ "chunked" ]
+    H.(get_multi prebuilt "transfer-encoding");
+  aesl "get_multi h k when mem h k = true"
+    [ "application/xml"; "text/html" ]
+    H.(get_multi prebuilt "accept")
+
+let hstr =
+  [
+    ("accept", "application/xml");
+    ("transfer-encoding", "chunked");
+    ("accept", "text/html");
+    ("content-length", "100");
+  ]
+
+let get_multi_concat_tests () =
+  let h1 = H.(add (add prebuilt "a" "a1") "a" "a2") in
+  aeso "get_multi_concat (init ()) k" None H.(get_multi_concat (init ()) "a");
+  aeso "get_multi_concat h k when mem h k = false" None
+    H.(get_multi_concat prebuilt "a");
+  aeso "get_multi_concat h k when mem h k = true"
+    (Some "application/xml,text/html")
+    H.(get_multi_concat prebuilt "accept");
+  aeso "get_multi_concat ~list_value_only:false h k when mem h k = true"
+    (Some "a1,a2")
+    H.(get_multi_concat h1 "a");
+  aeso "get_multi_concat ~list_value_only:true h k when mem h k = true"
+    (Some "a2")
+    H.(get_multi_concat ~list_value_only:true h1 "a")
+
+let map_tests () =
+  let a = ", a" in
+  aessl "map (fun _ v -> v) (init ())" []
+    H.(to_list (map (fun _k v -> v) (init ())));
+  aessl "map (fun _ v -> v) (init ())" (H.to_list prebuilt)
+    H.(to_list (map (fun _k v -> v) prebuilt));
+  aessl "map (fun _ v -> v ^ a ) (init ())"
+    [
+      ("accept", "application/xml, a");
+      ("transfer-encoding", "chunked, a");
+      ("accept", "text/html, a");
+      ("content-length", "100, a");
+    ]
+    H.(to_list (map (fun _k v -> v ^ a) prebuilt))
+
+let fold_tests () = ()
```

Comment:
Should we create an issue tracking the implementation of these two tests? Do you have in mind something specific already?

Reply:
I just added these missing tests.